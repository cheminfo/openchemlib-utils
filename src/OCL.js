let OCL;

export function getOCL() {
  if (!OCL) {
    throw new Error('OCL has to be initialized using initOCL(OCL)');
  }
  return OCL;
}
/**
 * 
 * @param {OCL} newOCL 
 * @param {object} [options={}]
 * @param {boolean} [options.keepExisting=false] Will keep current OCL if we initialize more than once
 */
export function initOCL(newOCL, options = {}) {
  const { keepExisting = false } = options;
  if (OCL && keepExisting) return;
  if (OCL && OCL !== newOCL) {
    throw new Error('OCL-utils was already initialized with a different OCL');
  }
  OCL = newOCL;
}
