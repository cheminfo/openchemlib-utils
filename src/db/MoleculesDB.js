import appendCSV from './utils/appendCSV';
import appendColor from './utils/appendColor';
import appendSDF from './utils/appendSDF';
import appendSmilesList from './utils/appendSmilesList';
import pushEntry from './utils/pushEntry';
import pushMoleculeInfo from './utils/pushMoleculeInfo';
import search from './utils/search';
/*
    this.db is an object with properties 'oclID' that has as value
    an object that contains the following properties:
    * molecule: an OCL molecule instance
    * index: OCL index used for substructure searching
    * properties: all the calculates properties
    * data: array containing free data associated with this molecule
  */

export class MoleculesDB {
  /**
   *
   * @param {OCL} [OCL] The openchemlib library
   * @param {object} [options={}]
   * @param {boolean} [options.computeProperties=false]
   */
  constructor(OCL, options = {}) {
    const { computeProperties = false } = options;
    this.OCL = OCL;
    this.db = {};
    this.statistics = null;
    this.computeProperties = computeProperties;
    this.searcher = new OCL.SSSearcherWithIndex();
  }

  /**
   * append to the current database a CSV file
   * @param {string|ArrayBuffer} csv - text file containing the comma separated value file
   * @param {object} [options={}]
   * @param {boolean} [options.header=true]
   * @param {boolean} [options.dynamicTyping=true]
   * @param {boolean} [options.skipEmptyLines=true]
   * @param {function} [options.onStep] call back to execute after each molecule
   */

  appendCSV(csv, options) {
    return appendCSV(this, csv, {
      computeProperties: this.computeProperties,
      ...options,
    });
  }

  /**
   * Append a SDF to the current database
   * @param {string|ArrayBuffer} sdf - text file containing the sdf
   * @param {object} [options={}]
   * @param {function} [options.onStep] call back to execute after each molecule
   * @returns {DB}
   */

  appendSDF(sdf, options) {
    return appendSDF(this, sdf, {
      computeProperties: this.computeProperties,
      ...options,
    });
  }

  /**
   * Append a SDF to the current database
   * @param {string|ArrayBuffer} smiles - text file containing a list of smiles
   * @param {object} [options={}]
   * @param {function} [options.onStep] call back to execute after each molecule
   * @returns {DB}
   */

  appendSmilesList(text, options) {
    return appendSmilesList(this, text, {
      computeProperties: this.computeProperties,
      ...options,
    });
  }

  /**
   * Add a molecule to the current database
   * @param {OCL.Molecule} molecule
   * @param {object} [data={}]
   * @param {object} [moleculeInfo={}] may contain precalculated index and mw
   */

  pushEntry(molecule, data, moleculeInfo) {
    pushEntry(this, molecule, data, moleculeInfo);
  }

  /**
   * Add an entry in the database
   * @param {object} moleculeInfo - a molecule as a JSON that may contain the following properties: molfile, smiles, idCode, mf, index
   * @param {object} [data={}]
   */

  pushMoleculeInfo(moleculeInfo, data) {
    return pushMoleculeInfo(this, moleculeInfo, data);
  }

  /**
   * Search in a MoleculesDB
   * Inside the database all the same molecules are group together
   * @param {string|OCL.Molecule} [query] smiles, molfile, oclCode or instance of Molecule to look for
   * @param {object} [options={}]
   * @param {string} [options.format='idCode'] - query is in the format 'smiles', 'oclid' or 'molfile'
   * @param {string} [options.mode='substructure'] - search by 'substructure', 'exact' or 'similarity'
   * @param {boolean} [options.flattenResult=true] - The database group the data for the same product. This allows to flatten the result
   * @param {boolean} [options.keepMolecule=false] - keep the OCL.Molecule object in the result
   * @param {number} [options.limit=Number.MAX_SAFE_INTEGER] - maximal number of result
   * @return {Array} array of object of the type {(molecule), idCode, data, properties}
   */
  search(query, options) {
    return search(this, query, options);
  }

  /**
   * Returns an array with the current database
   * @returns
   */
  getDB() {
    return Object.keys(this.db).map((key) => this.db[key]);
  }

  /**
   * Append the property `data.color` to each entry based on a data or property label
   * {object} [options={}]
   * {string} [options.dataLabel] name of the property from `data` to use
   * {string} [options.propertyLabel] name of the property from `properties` to use
   * {number} [options.colorLabel='color'] name of the property to add in data that will contain the color
   * {number} [options.minValue]
   * {number} [options.maxValue]
   * {number} [options.minHue=0]
   * {number} [options.maxHue=360]
   * {number} [options.saturation=65] percent of color saturation
   * {number} [options.lightness=65] percent of color lightness
   */
  appendColor(options) {
    appendColor(this, options);
  }
}
