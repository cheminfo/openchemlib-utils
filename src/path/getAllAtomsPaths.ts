import { Molecule } from 'openchemlib';

export interface GetAllAtomsPathsOptions {
  /**
   * Maximum path length to consider. All the paths from 0 to maxPathLength will be considered
   * @default 5
   */
  maxPathLength?: number;
}

interface AtomPath {
  path: number[];
  distance: number;
}

/**
 * We need to create an array of atoms
 * that contains an array of pathLength
 * that contains an array of object
 * @param molecule
 * @param options
 * @returns
 */
export function getAllAtomsPaths(
  molecule: Molecule,
  options: GetAllAtomsPathsOptions = {},
) {
  const { maxPathLength = 5 } = options;
  const allAtomsPaths = [];
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    const oneAtomPaths: AtomPath[][] = [];
    allAtomsPaths.push(oneAtomPaths);

    let atomPaths: AtomPath[] = [];
    atomPaths.push({ path: [i], distance: 0 });
    oneAtomPaths.push(atomPaths);

    let nextIndexes = [0];
    let nextAtoms = [i];

    for (let sphere = 1; sphere <= maxPathLength; sphere++) {
      atomPaths = [];
      oneAtomPaths.push(atomPaths);
      const currentIndexes = nextIndexes;
      const currentAtoms = nextAtoms;
      nextIndexes = [];
      nextAtoms = [];

      for (let i = 0; i < currentIndexes.length; i++) {
        const atom = currentAtoms[i];
        const index = currentIndexes[i];

        const previousPath = oneAtomPaths[sphere - 1][index].path;
        for (let conn = 0; conn < molecule.getConnAtoms(atom); conn++) {
          const connectedAtom = molecule.getConnAtom(atom, conn);
          if (previousPath.includes(connectedAtom)) continue;
          nextIndexes.push(atomPaths.length);
          nextAtoms.push(connectedAtom);
          atomPaths.push({
            path: [...previousPath, connectedAtom],
            distance: sphere,
          });
        }
      }
    }
  }
  return allAtomsPaths;
}
