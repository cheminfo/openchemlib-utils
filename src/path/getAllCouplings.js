import { tagAtom } from '../util/tagAtom';
import { makeRacemic } from '../util/makeRacemic';

/**
 * Returns an array of all the different atom diaIDs that are connected
 * {OCL}
 * {OCL.Molecule} molecule
 * {object} [options={}]
 * {string} [options.fromLabel='H']
 * {string} [options.toLabel='H']
 * {number} [options.minLength=1]
 * {number} [options.maxLength=4]
 */

function getAllCouplings(OCL, molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;
  let paths = molecule.getAllPaths({
    fromLabel,
    toLabel,
    minLength,
    maxLength,
  });

  const minSphereSize = 0;
  const maxSphereSize = 2;

  let fragment = new OCL.Molecule(0, 0);
  for (let path of paths) {
    path.info = [];
    for (let fromTo of path.fromTo) {
      let atoms = [];
      molecule.getPath(atoms, fromTo[0], fromTo[1], path.pathLength);
      let torsion;
      if (atoms.length === 4) {
        torsion = molecule.calculateTorsion(atoms);
      }
      path.info.push({
        atoms,
        torsion,
      });

      if (!path.code) {
        path.code = [];

        let tmpMolecule = molecule.getCompactCopy();
        makeRacemic(OCL, tmpMolecule);
        tagAtom(OCL, tmpMolecule, atoms[0]);
        tagAtom(OCL, tmpMolecule, atoms[atoms.length - 1]);

        let atomMask = new Array(tmpMolecule.getAllAtoms()).fill(false);
        let atomList = [];
        let max = 0;
        let min = 0;
        for (let sphere = 0; sphere <= maxSphereSize; sphere++) {
          if (max === 0) {
            for (let atom of atoms) {
              atomMask[atom] = true;
              atomList.push(atom);
              max++;
            }
          } else {
            let newMax = max;
            for (let i = min; i < max; i++) {
              let atom = atomList[i];
              for (let j = 0; j < tmpMolecule.getConnAtoms(atom); j++) {
                let connAtom = tmpMolecule.getConnAtom(atom, j);
                atomMask[connAtom] = true;
                atomList[newMax++] = connAtom;
              }
            }
            min = max;
            max = newMax;
          }
          tmpMolecule.copyMoleculeByAtoms(fragment, atomMask, true, null);
          if (sphere >= minSphereSize) {
            path.code.push(
              fragment.getCanonizedIDCode(
                OCL.Molecule.CANONIZER_ENCODE_ATOM_CUSTOM_LABELS,
              ),
            );
          }
        }
      }
    }
  }
  return paths;
}

module.exports = getAllCouplings;
