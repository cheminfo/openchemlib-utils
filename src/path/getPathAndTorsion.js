/**
 * Calculates the path between 2 atoms
 * @param {import('openchemlib').Molecule} molecule
 * @param {number} from - index of the first atom
 * @param {number} to - index of the end atom
 * @param {number} maxLength - maximal length of the path
 */
export function getPathAndTorsion(molecule, from, to, maxLength) {
  let originalAtoms = []; // path before renumbering
  molecule.getPath(originalAtoms, from, to, maxLength + 1);
  let torsion;
  if (originalAtoms.length === 4) {
    torsion = molecule.calculateTorsion(originalAtoms);
  }

  return {
    atoms: originalAtoms,
    from,
    to,
    torsion,
    length: originalAtoms.length - 1,
  };
}
