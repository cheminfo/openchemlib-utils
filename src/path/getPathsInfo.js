import { getHoseCodesForPath } from '../hose/getHoseCodesForPath';
import { getAtomsInfo } from '../util/getAtomsInfo';
import { getConnectivityMatrix } from '../util/getConnectivityMatrix';

let fragment;

/**
 *
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.minLength=1]
 * @param {string} [opions.maxLength=4]

 */
export function getPathsInfo(molecule, options = {}) {
  const {
    fromLabel = 'H',
    toLabel = 'H',
    minLength = 1,
    maxLength = 4,
  } = options;

  const OCL = molecule.getOCL();

  if (!fragment) {
    fragment = new OCL.Molecule(0, 0);
  }

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  // we need to find all the atoms 'fromLabel' and 'toLabel'
  let atomsInfo = getAtomsInfo(molecule);

  let pathLengthMatrix = getConnectivityMatrix(molecule, {
    pathLength: true,
  });

  for (let from = 0; from < molecule.getAllAtoms(); from++) {
    atomsInfo[from].paths = [];
    for (let to = 0; to < molecule.getAllAtoms(); to++) {
      if (from !== to) {
        if (molecule.getAtomicNo(from) === fromAtomicNumber) {
          if (molecule.getAtomicNo(to) === toAtomicNumber) {
            let pathLength = pathLengthMatrix[from][to];
            if (pathLength >= minLength && pathLength <= maxLength) {
              atomsInfo[from].paths.push(
                getHoseCodesForPath(
                  molecule,
                  from,
                  to,
                  pathLength,
                  atomsInfo[to].oclID,
                ),
              );
            }
          }
        }
      }
    }
  }

  return atomsInfo;
}
