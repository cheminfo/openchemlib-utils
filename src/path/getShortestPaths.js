/**
 * Get the shortest path between each pair of atoms in the molecule
 * @param {OCL.Molecule} molecule
 * @param {object} [options={}]
 * @param {string} [opions.fromLabel='H']
 * @param {string} [opions.toLabel='H']
 * @param {string} [opions.maxLength=4]
 * @returns {Array<Array>} A matrix containing on each cell (i,j) the shortest path from atom i to atom j
 */
export function getShortestPaths(molecule, options = {}) {
  const OCL = molecule.getOCL();
  const { fromLabel = '', toLabel = '', maxLength = 3 } = options;

  let fromAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(fromLabel);
  let toAtomicNumber = OCL.Molecule.getAtomicNoFromLabel(toLabel);

  const nbAtoms = molecule.getAllAtoms();
  let allShortestPaths = new Array(nbAtoms);
  for (let i = 0; i < nbAtoms; i++) {
    allShortestPaths[i] = new Array(nbAtoms);
  }

  for (let from = 0; from < nbAtoms; from++) {
    allShortestPaths[from][from] = [from];
    for (let to = from + 1; to < nbAtoms; to++) {
      if (
        (fromAtomicNumber === 0 ||
          molecule.getAtomicNo(from) === fromAtomicNumber) &&
        (toAtomicNumber === 0 || molecule.getAtomicNo(to) === toAtomicNumber)
      ) {
        let path = [];
        molecule.getPath(path, from, to, maxLength);
        if (path.length) {
          allShortestPaths[from][to] = path.slice();
          allShortestPaths[to][from] = path.reverse();
        } else {
          allShortestPaths[from][to] = null;
          allShortestPaths[to][from] = null;
        }
      } else {
        allShortestPaths[from][to] = null;
        allShortestPaths[to][from] = null;
      }
    }
  }

  return allShortestPaths;
}
