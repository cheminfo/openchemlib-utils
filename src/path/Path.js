import { getPathsInfo } from './getPathsInfo';

export const Path = {
  getPathsInfo,
};
