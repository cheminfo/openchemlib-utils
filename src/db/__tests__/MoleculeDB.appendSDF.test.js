import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import OCL from 'openchemlib';
import { expect, it, describe } from 'vitest';

import { MoleculesDB } from '../MoleculesDB';

describe('appendSDF', () => {
  const sdf = readFileSync(join(__dirname, './data/data.sdf'));
  it('should parse all molecules from SDF', async () => {
    const moleculesDB = new MoleculesDB(OCL);
    await moleculesDB.appendSDF(sdf);
    expect(Object.keys(moleculesDB.db)).toHaveLength(10);
    const db = moleculesDB.getDB();
    expect(db).toHaveLength(10);
    expect(db.filter((entry) => entry.properties)).toHaveLength(10);
  });

  it('should call step for each molecule', () => {
    const moleculesDB = new MoleculesDB(OCL);
    let called = 0;

    function onStep(current, total) {
      expect(current).toBe(++called);
      expect(total).toBe(20);
    }

    return moleculesDB.appendSDF(sdf, { onStep }).then(() => {
      expect(called).toBe(20);
    });
  });

  it('should compute properties', async () => {
    const moleculesDB = new MoleculesDB(OCL, { computeProperties: true });
    await moleculesDB.appendSDF(sdf);
    const firstEntry = moleculesDB.getDB()[0];
    expect(firstEntry).toHaveProperty('properties');
    expect(firstEntry.properties.mf).toBeDefined();
    expect(firstEntry.properties.logP).toBeDefined();
    expect(firstEntry.properties.polarSurfaceArea).toBeDefined();
  });

  it('stats should be in incremental', async () => {
    const moleculesDB = new MoleculesDB(OCL, { computeProperties: true });
    await moleculesDB.appendSDF(sdf);
    expect(moleculesDB.statistics).toMatchSnapshot();
    expect(
      moleculesDB.statistics.data.map((datum) => datum.counter),
    ).toStrictEqual([20, 20, 20, 19, 19, 9, 19]);
    const sdfTest = readFileSync(join(__dirname, './data/test.sdf'));
    await moleculesDB.appendSDF(sdfTest);
    expect(
      moleculesDB.statistics.data.map((datum) => datum.counter),
    ).toStrictEqual([
      28, 20, 20, 19, 19, 9, 19, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 2, 2, 2, 2,
    ]);
    expect(moleculesDB.statistics).toMatchSnapshot();
    moleculesDB.clear();
    await moleculesDB.appendSDF(sdf);
    expect(
      moleculesDB.statistics.data.map((datum) => datum.counter),
    ).toStrictEqual([20, 20, 20, 19, 19, 9, 19]);
  });
});
