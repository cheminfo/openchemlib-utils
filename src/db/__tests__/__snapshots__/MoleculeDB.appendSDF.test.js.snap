// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`appendSDF > stats should be in incremental 1`] = `
{
  "calculated": [
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "mw",
      "maxValue": 402.48519999999996,
      "minValue": 30.02588,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "em",
      "maxValue": 402.20424,
      "minValue": 30.010565,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": false,
      "kind": "string",
      "label": "mf",
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "acceptorCount",
      "maxValue": 6,
      "minValue": 1,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "donorCount",
      "maxValue": 3,
      "minValue": 0,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "logP",
      "maxValue": 7.642299979925156,
      "minValue": -1.71840001642704,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "logS",
      "maxValue": 0.022000011056661606,
      "minValue": -5.927000029012561,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "polarSurfaceArea",
      "maxValue": 97.73999786376953,
      "minValue": 17.06999969482422,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "rotatableBondCount",
      "maxValue": 5,
      "minValue": 0,
    },
    {
      "always": true,
      "counter": 10,
      "isNumeric": true,
      "kind": "number",
      "label": "stereoCenterCount",
      "maxValue": 8,
      "minValue": 0,
    },
  ],
  "data": [
    {
      "always": true,
      "counter": 20,
      "isNumber": false,
      "kind": "string",
      "label": "molfile",
    },
    {
      "always": true,
      "counter": 20,
      "isNumber": false,
      "kind": "string",
      "label": "ID",
    },
    {
      "always": true,
      "counter": 20,
      "isNumber": false,
      "kind": "string",
      "label": "rn",
    },
    {
      "always": false,
      "counter": 19,
      "isNumber": false,
      "kind": "string",
      "label": "mf",
    },
    {
      "always": false,
      "counter": 19,
      "isNumber": false,
      "isNumeric": true,
      "kind": "number",
      "label": "mw",
      "maxValue": 402.49,
      "minValue": 30.03,
    },
    {
      "always": false,
      "counter": 9,
      "isNumber": false,
      "kind": "string",
      "label": "bp",
    },
    {
      "always": false,
      "counter": 19,
      "isNumber": false,
      "kind": "string",
      "label": "mp",
    },
  ],
}
`;

exports[`appendSDF > stats should be in incremental 2`] = `
{
  "calculated": [
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "mw",
      "maxValue": 402.48519999999996,
      "minValue": 30.02588,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "em",
      "maxValue": 402.20424,
      "minValue": 30.010565,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": false,
      "kind": "string",
      "label": "mf",
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "acceptorCount",
      "maxValue": 9,
      "minValue": 1,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "donorCount",
      "maxValue": 4,
      "minValue": 0,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "logP",
      "maxValue": 7.642299979925156,
      "minValue": -1.71840001642704,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "logS",
      "maxValue": 0.022000011056661606,
      "minValue": -5.927000029012561,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "polarSurfaceArea",
      "maxValue": 139.53999948501587,
      "minValue": 17.06999969482422,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "rotatableBondCount",
      "maxValue": 9,
      "minValue": 0,
    },
    {
      "always": true,
      "counter": 18,
      "isNumeric": true,
      "kind": "number",
      "label": "stereoCenterCount",
      "maxValue": 8,
      "minValue": 0,
    },
  ],
  "data": [
    {
      "always": true,
      "counter": 28,
      "isNumber": false,
      "kind": "string",
      "label": "molfile",
    },
    {
      "always": false,
      "counter": 20,
      "isNumber": false,
      "kind": "string",
      "label": "ID",
    },
    {
      "always": false,
      "counter": 20,
      "isNumber": false,
      "kind": "string",
      "label": "rn",
    },
    {
      "always": false,
      "counter": 19,
      "isNumber": false,
      "kind": "string",
      "label": "mf",
    },
    {
      "always": false,
      "counter": 19,
      "isNumber": false,
      "isNumeric": true,
      "kind": "number",
      "label": "mw",
      "maxValue": 402.49,
      "minValue": 30.03,
    },
    {
      "always": false,
      "counter": 9,
      "isNumber": false,
      "kind": "string",
      "label": "bp",
    },
    {
      "always": false,
      "counter": 19,
      "isNumber": false,
      "kind": "string",
      "label": "mp",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Catalogue Number",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Chemical Name",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "CAS Number",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Molecular Formula",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "isNumeric": true,
      "kind": "number",
      "label": "Molecular Weight",
      "maxValue": 321.33,
      "minValue": 222.72,
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Quantity1",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Price1",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Quantity2",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Price2",
    },
    {
      "always": false,
      "counter": 8,
      "isNumber": false,
      "kind": "string",
      "label": "Category",
    },
    {
      "always": false,
      "counter": 2,
      "isNumber": false,
      "kind": "string",
      "label": "Quantity3",
    },
    {
      "always": false,
      "counter": 2,
      "isNumber": false,
      "kind": "string",
      "label": "Price3",
    },
    {
      "always": false,
      "counter": 2,
      "isNumber": false,
      "kind": "string",
      "label": "Quantity4",
    },
    {
      "always": false,
      "counter": 2,
      "isNumber": false,
      "kind": "string",
      "label": "Price4",
    },
  ],
}
`;
